docs/proposals/29-autoscaling-storage-volumes.md

```diff
@@ -0,0 +1,335 @@
+  - Autoscaling is inactive by default.
+  - Ability to selectively activate autoscaling per PVC.
+  - Ability to trigger scaling based on relative volume utilisation ratio (`<used bytes> / <capacity bytes>`). Trigger
+    threshold configurable via code.
+  - Ability to configure absolute upper scaling limit per PVC.
+  - Autoscaling causes minimal interruption to scaled workloads.
+- Autoscaling ability activated for observability volumes (`prometheus` and `vali`) in seeds'
+  `garden` namespace, and in shoot namespaces.
+- Absolute upper scaling limit configured on each observability volume.
+- Reduce the initial size of a volume, when autoscaling is activated for that volume, if such reduction is aligned with
+  general Gardener goals (cost, reliability).
+- Some metrics, which represent an overview of the PVC autoscaling state per seed, available through one or more of the seed
+  system's Prometheus instances.
+- Provide operational documentation for Gardener Operators.
+- Ability to enable/disable the feature per seed, via feature gate.
+
+### Non-Goals
+- Ability to activate autoscaling on a group of PVCs, other than specifying each PVC individually. Activate by selector
+  or namespace is not supported.
+- Ability to selectively deactivate autoscaling, e.g. by PVC selector or namespace.
+- Ability to scale a PVC if the storage class does not support resizing.
+- Ability to deactivate autoscaling for any of the observability volumes specified in the 'Goals' section.
+- Ability to scale volumes in shoot clusters.
+
+### Notes:
+- Shrinking a PVC is not supported. A limitation imposed by the underlying infrastructure, PVC size can only be
+  increased, and not decreased.
+- The newly introduced auto-scaling mode is predicated on the underlying infrastructure supporting volume resizing.
+  This ability is expressed through the K8s StorageClass type. The respective StorageClass must have
+  `allowVolumeExpansion: true`. For details, see the K8s volume expansion documentation.
+  
+## Proposal
+### PVC Autoscaler - Runtime Structure
+The overall runtime structure of the component is outlined in _Fig.1_:
+
+![01-runtime_structure.png](./assets/gep29-01-runtime_structure.png)
+
+_Fig.1: Runtime structure_
+
+For the role of a storage scaling controller, the existing, recently implemented Gardener PVC autoscaler application
+[[1]], hereafter referred to as `pvc-autoscaler`, is used. It is
+integrated as a seed system component, and runs as a deployment in the seed's `garden` namespace.
+Its primary driving signal is the PVC metrics from the seed's cache Prometheus instance, which it
+periodically examines, and if capacity is found to be near exhaustion, `pvc-autoscaler` takes action by updating the
+PVC's storage request.
+
+In this initial iteration, `pvc-autoscaler` scales two categories of `vali` and `prometheus` volumes:
+those in shoot namespaces, and those in the seed's `garden` namespace.
```

Comment:
Yes. I considered it, but I think it's not worth doing at this time. There are few observability volumes in the garden cluster and, at least in productive gardens, the utilisation doesn't vary greatly. So preallocating a bit of extra space works well enough for the garden.

Once the feature is mature, I expect it will be worth adding to the garden cluster. But I expect that it will still have a relatively low priority because of the low volume count there.

Reply:
To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.